### PR TITLE
executor: make alias identifier maximum length compatible with MySQL

### LINKED ISSUE #10513: Inconsisitent Behaviour with Mysql When Col Name is too Long
## Bug Report

When col name is too long, mysql will cut a part of it. But inTiDB, col name will not be cut

1. What did you do?

```sql
SELECT COALESCE(0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0);
```

2. What did you expect to see?

tidb will output the same with mysql.

3. What did you see instead?

mysql cut the col name, but tidb not.

As follow. left is mysql and right is tidb.

big col name

the col name mysql output:

```sql
COALESCE(0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
```

but col name tidb output:

```sql
COALESCE(0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
Release Version: v3.0.0-beta-111-g266ff4b6f
Git Commit Hash: 266ff4b6f37e721cf3d777a4e1d94b692f235d60
Git Branch: master
UTC Build Time: 2019-02-22 10:50:48
GoVersion: go version go1.11.2 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
```

Mysql version is 5.7

### PR DESCRIPTION
**Pull request by miraclesu**
### What problem does this PR solve? 
Fix issue #10513 .

### What is changed and how it works?
Truncate alias if the length of alias is greater than `mysql.MaxAliasIdentifierLen`.

### Check List 

Tests 

 - Unit test

Code changes

N/A

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: executor/adapter.go [hunk 1]
```diff
@@ -89,6 +89,15 @@ func schema2ResultFields(schema *expression.Schema, defaultDB string) (rfs []*as
 				Name:      origColName,
 			},
 		}
+		// This is for compatibility.
+		if len(rf.ColumnAsName.O) > mysql.MaxAliasIdentifierLen {
```
Comment (lamxTyler) [inline executor/adapter.go]: Seems that the length is measured in characters, not bytes, so maybe we should calculate the run count to determine if it is needed to truncate.

Comment (miraclesu) [reply executor/adapter.go]: According to the [MySQL documentation](https://dev.mysql.com/doc/refman/5.7/en/identifiers.html)
> Maximum Length (characters)

The length is measured in characters.

Comment (zz-jason) [reply executor/adapter.go]: "run count" -> "rune count". As you can see in the mysql document, `mysql.MaxAliasIdentifierLen` limits the character count, not the byte count. For example, "中文" only has 2 characters, but it has 6 bytes.

Comment (miraclesu) [reply executor/adapter.go]: I've tested in MySQL, the result as follow, did I miss something?
image

Comment (zz-jason) [reply executor/adapter.go]: :joy: You should test the behavior of Golang `len()` function, not MySQL builtin function `LENGTH()`. They are different. 

```go
package main

import "fmt"

func main() {
    fmt.Println(len("中文")) // result is 6
    fmt.Println(len([]rune("中文"))) // result is 2
}
```

Comment (miraclesu) [reply executor/adapter.go]: @zz-jason Shown in the picture above, MySQL alias truncated the Chinese character.
If MySQL alias length is measured in characters(i.e. Chinese character), the character will never be truncated, right?
Is that can prove MySQL alias length is measured in byte, not characters?
So in golang code, we should use len, not len rune, right?

Comment (zz-jason) [reply executor/adapter.go]: Yeah. Seems MySQL document is incorrect compared with its actual behavior.

**Pull request merged**